### PR TITLE
SUPPORT-139 provider: add a chia-aware bid price script

### PR DESCRIPTION
**Pull request by ghost**
This PR contains 3 commits for better history tracking.

To use Andrew's script just set this parameter:

```
--set bidpricescript="$(cat ./charts/akash-provider/scripts/price_script_chia.sh | openssl base64 -A)"
```

----

## Testing this script directly

You can test this script quickly this way:

**1 replica:**

```
cd charts/akash-provider/scripts
./price_script_chia.sh <<< '[{"memory":1073741824,"cpu":1000,"storage":[{"class":"ephemeral","size":1073741824}],"count":1,"endpoint_quantity":2}]'
```

**3 replicas:**

```
cd charts/akash-provider/scripts
./price_script_chia.sh <<< '[{"memory":1073741824,"cpu":1000,"storage":[{"class":"ephemeral","size":1073741824}],"count":1,"endpoint_quantity":1},{"memory":536870912,"cpu":100,"storage":[{"class":"ephemeral","size":536870912}],"count":1,"endpoint_quantity":0},{"memory":536870912,"cpu":100,"storage":[{"class":"ephemeral","size":536870912}],"count":1,"endpoint_quantity":0}]'
```

> add `bash -x ./price_script_chia.sh <<< ...` for more verbosity.

### CODE REVIEW

File: charts/akash-provider/scripts/price_script_chia.sh [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+cpu_total=$(echo "$data_in" | jq 'map(.cpu * .count) | add')
+memory_total=$(echo "$data_in" | jq 'map(.memory * .count) | add')
+storage_total=$(echo "$data_in" |  python3 -c "import json;import sys;data=json.load(sys.stdin);data = [(x['count'], x['storage']) for x in data];data = [cnt*sum(z['size'] for z in y) for cnt,y in data]; data= sum(data);json.dump(data,sys.stdout)")
+
+cpu_total_threads=$(echo $cpu_total | awk '{print $1/1000}')
+memory_gb=$(echo $memory_total | awk '{print $1/1024/1024/1024}')
+hd_gb=$(echo $storage_total | awk '{print $1/1024/1024/1024}')
+
+
+usd_per_akt=$(curl -s -X GET "https://api.coingecko.com/api/v3/coins/akash-network/tickers" -H  "accept: application/json" | jq '.tickers[] | select(.market.name == "Osmosis").converted_last.usd' | head -n1)
```
Comment (sacreman) [inline charts/akash-provider/scripts/price_script_chia.sh]: Maybe add `--connection-timeout 5`

Comment (ghost) [reply charts/akash-provider/scripts/price_script_chia.sh]: Done.

**Comment by ghost** [comment]
Looks like coingecko API is now behind CloudFlare and it wants Captcha now :/

```
# curl -s -X GET https://api.coingecko.com/api/v3/coins/akash-network/tickers -H 'accept: application/json'
...
<title>Please Wait... | Cloudflare</title>
...
```

**Comment by ghost** [comment]
Changed CoinGecko API to Osmosis.

**Comment by ghost** [comment]
Removed python3 dependency from the original script.

File: charts/akash-provider/scripts/price_script_chia.sh [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+
+# cache AKT price for 60 minutes to reduce the API pressure as well as to slightly accelerate the bidding (+5s)
+CACHE_FILE=/tmp/aktprice.cache
+if ! test $(find $CACHE_FILE -mmin -60 2>/dev/null); then
+  ## cache expired
+  usd_per_akt=$(curl -s --connect-timeout 5 -X GET 'https://api-osmosis.imperator.co/tokens/v2/price/AKT' -H 'accept: application/json' | jq -r '.price')
+  # TODO: add an alternative API to avoid SPOF.
+
+  # update the cache only when API returns a result.
```
Comment (boz) [inline charts/akash-provider/scripts/price_script_chia.sh:29]: may want to check that it has the proper numeric format here.

File: charts/akash-provider/scripts/price_script_chia.sh [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+chia_bladebit_memory_max=512
+
+chia_bladebit_storage=715
+chia_bladebit_storage_max=3200
+
+if (( $(echo "$memory_gb >= $chia_bladebit_memory" | bc -l) && \
+      $(echo "$cpu_total_threads >= $chia_bladebit_cpu" | bc -l) )); then
+#Bladebit detected
+TARGET_CPU="20"
+total_cost_usd_target=$(bc -l <<<"($cpu_total_threads * $TARGET_CPU)")
```
Comment (boz) [inline charts/akash-provider/scripts/price_script_chia.sh]: mind properly indenting these?

File: charts/akash-provider/templates/configmap-bidscripts.yaml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "provider.fullname" . }}-bidscripts
+  namespace: {{ .Release.Namespace }}
+data:
+  price_script_chia.sh: |
+{{ tpl (.Files.Get "scripts/price_script_chia.sh") . | indent 4 }}
```
Comment (boz) [inline charts/akash-provider/templates/configmap-bidscripts.yaml]: it'd be nice to parameterize this path.

**Review by boz — APPROVED** [review]
Looks great, thanks!

File: charts/akash-provider/scripts/price_script_chia.sh (hunk 2 above)
Comment (ghost) [reply charts/akash-provider/scripts/price_script_chia.sh:29]: addressed.
additionally added a simple range check.

File: charts/akash-provider/scripts/price_script_chia.sh (hunk 3 above)
Comment (ghost) [reply charts/akash-provider/scripts/price_script_chia.sh]: addressed.

File: charts/akash-provider/templates/configmap-bidscripts.yaml (hunk 4 above)
Comment (ghost) [reply charts/akash-provider/templates/configmap-bidscripts.yaml]: addressed.

akash-provider will always run `/bidscripts/price_script.sh` [hardcoded] (inside the pod, this is hidden from the user).

To use bid price script, the user only needs to pass this argument:

```
--set bidpricescript="$(cat ./charts/akash-provider/scripts/price_script_chia.sh | openssl base64 -A)"
```

Comment (boz) [reply charts/akash-provider/templates/configmap-bidscripts.yaml]: we should just pass a file path in.  it can be addressed later.

Comment (ghost) [reply charts/akash-provider/templates/configmap-bidscripts.yaml]: If you think that we should pass a file path for the akash provider running in the pod itself (`AKASH_BID_PRICE_SCRIPT_PATH`), then what is the use-case?
I think the user should only care of passing the bid price script itself, as ConfigMap is automatically making sure the path is always correct.

**Pull request merged**